sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/AttributeSet.scala

```diff
@@ -105,10 +105,12 @@ class AttributeSet private (private val baseSet: mutable.LinkedHashSet[Attribute
    */
   def --(other: Iterable[NamedExpression]): AttributeSet = {
     other match {
+      // SPARK-32755: `--` method behave differently under scala 2.12 and 2.13,
+      // use a Scala 2.12 based code to maintains the insertion order in Scala 2.13
       case otherSet: AttributeSet =>
-        new AttributeSet(baseSet -- otherSet.baseSet)
+        new AttributeSet(baseSet.clone() --= otherSet.baseSet)
```

Comment:
In Scala 2.12 the implementation of `--` is `  override def --(xs: GenTraversableOnce[A]): This = clone() --= xs.seq
`.

Reply:
`diff` method is defined as `Computes the difference of this set and another set`, I haven't tested overall behavior of `diff`, but I think the current changes don't look complicated either

Reply:
diff returns a set, so it can potentially mess up the ordering.

Reply:
![image](https://user-images.githubusercontent.com/1475305/92564402-ae8c9680-f2ab-11ea-9425-350c0d5f5b03.png)
@hvanhovell Actually, `diff` returns `LinkedHashSet[AttributeEquals]` here...

Reply:
In Scala 2.13 
```
def diff(that: collection.Set[A]): C =
    toIterable.foldLeft(empty)((result, elem) => if (that contains elem) result else result += elem)
```

In Scala 2.12
```
def diff(that: GenSet[A]): This = this -- that
```

It looks like they're maintains ordered, but the input parameter needs to be a `Set`, so 

```
other.map(a => new AttributeEquals(a.toAttribute))
```

part need call `toSet`。。。

Reply:
Ok, then diff looks better.

Reply:
If we can avoid the conversion it would be nice.

Reply:
This conversion seems inevitable if use `diff` :(

Reply:
Let's use the current method then :).

Reply:
ok ~